src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -1313,6 +1261,76 @@ abstract class ClassfileParser {
       sym setInfo createFromClonedSymbols(alias.initialize.typeParams, alias.tpe)(typeFun)
     }
   }
+  final class JavaSignatureTypeCompleter(name: Name, jflags: JavaAccFlags, index: Int) extends LazyType {
+    override def complete(sym: symbolTable.Symbol): Unit = {
+      var info = pool.getType(sym, index)
+      if (name == nme.CONSTRUCTOR)
+        info match {
+          case MethodType(params, restpe) =>
+            // if this is a non-static inner class, remove the explicit outer parameter
+            val paramsNoOuter = innerClasses getEntry currentClass match {
+              case Some(entry) if !entry.jflags.isStatic =>
+                /* About `clazz.owner.hasPackageFlag` below: scala/bug#5957
+                 * For every nested java class A$B, there are two symbols in the scala compiler.
+                 *  1. created by SymbolLoader, because of the existence of the A$B.class file, owner: package
+                 *  2. created by ClassfileParser of A when reading the inner classes, owner: A
+                 * If symbol 1 gets completed (e.g. because the compiled source mentions `A$B`, not `A#B`), the
+                 * ClassfileParser for 1 executes, and clazz.owner is the package.
+                 */
+                assert(params.head.tpe.typeSymbol == clazz.owner || clazz.owner.hasPackageFlag, params.head.tpe.typeSymbol + ": " + clazz.owner)
+                params.tail
+              case _ =>
+                params
+            }
+            val newParams = paramsNoOuter match {
+              case (init :+ tail) if jflags.isSynthetic =>
+                // scala/bug#7455 strip trailing dummy argument ("access constructor tag") from synthetic constructors which
+                // are added when an inner class needs to access a private constructor.
+                init
+              case _ =>
+                paramsNoOuter
+            }
+
+            info = MethodType(newParams, clazz.tpe)
+        }
+      // Note: the info may be overwritten later with a generic signature
+      // parsed from SignatureATTR
+      sym.setInfo(if (jflags.isVarargs) arrayToRepeated(info) else info)
+      addParamNames(sym)
+    }
+  }
+
+  private def addParamNames(sym: Symbol): Unit = {
+    sym.getAndRemoveAttachment[ParamNames] match {
+      case Some(ParamNames(paramNames)) =>
+        if (sym.hasRawInfo && sym.isMethod) {
+          val params = sym.rawInfo.params
+          val paramNames1 = paramNames.takeRight(params.length)
```

Comment:
add a comment why. also the [spec](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.24) does not reeeally mandate the number of parameter names to match the number of parameters, so things would be shifted here.

Since the `complete` method (which drops the outer) runs right before `addParamNames`, we can just pass the fact as a parameter to `addParamNames`, no?

Reply:
Agreed, I'll rework this. We should probably grab both the descriptor and the signature (if it exists) and the param names (if they exist) in `parseAttributes`, and pass all of that into the completer. The completer can then figure out which param names to use.

The param names are specced to correspond to the descriptor signature (perhaps only a prefix of it, though).